### PR TITLE
Refactor for modularity

### LINKED ISSUE #58: Refactor for modularity
**Feature Overview**
What was once a small script meant to give developers a head start on importing their Botmock projects into Dialogflow is now facing increased demands along the lines of reliability, consistency and correctness.

There are aspects of the current implementation that are unmaintainable from this angle. Some of these aspects are:
- much of the implementation of the file writing process is happening directly in `index.ts`
- the file writing implementation is coupled to global variables
- there is no reliance on the canonical Botmock SDK
- there is no clear way to write unit tests for the functions that are interwoven with one another
- there is no clear way to write integration tests for the functions that perform related operations yet are separated and relying on the same global variables

**Describe the solution you'd like**
The repo should undergo a refactor for the sake of future maintainability, ease of testing, and assurance of correctness.

### PR DESCRIPTION
**Pull request by nonnontrivial**
This PR resolves #58 

## Includes
- [x] Unit tests for file writing
- [x] Unit tests for project asset fetching
- [x] Integration tests
- [x] Variables in utterances are correctly split
- [x] https://github.com/Botmock/botmock-dialogflow-export/issues/36 is resolved
- [x] Intent files are written
- [x] Input context is correctly implemented
- [x] https://github.com/Botmock/botmock-dialogflow-export/issues/46 is resolved
- [x] Output context is correctly implemented
- [x] https://github.com/Botmock/botmock-dialogflow-export/issues/60 is resolved
- [x] Entity files are written
- [x] Meta files are written
- [x] Missing welcome intent is detected and written with correct message content
- [x] Custom entities are correctly referenced in utterances
- [x] Presence of slot filling for intent is reflected in `parameters` field
- [x] Resulting output imports without error once compressed
- [x] https://github.com/Botmock/botmock-dialogflow-export/issues/56 is resolved
- [x] Docs are updated to reflect latest changes
- [x] Variable signs in response texts are correct

### CODE REVIEW

File: lib/providers/index.ts [hunk 1]
```diff
@@ -1,52 +1,63 @@
...
+        );
     }
-    if (type.endsWith("button") || type.endsWith("generic")) {
-      method = "card";
-    }
-    if (!method) {
+    const platform = this.platform.constructor.name.toLowerCase();
+    if (!methodToCallOnClass) {
       return {
         type: 4,
```
Comment (obaid) [inline lib/providers/index.ts]: What does `type` refer to here?

File: lib/providers/platforms/facebook.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+export default class Facebook {
+  /**
+   * 
+   * @param data 
+   */
+  public text(data: any) {
+    return { type: 0, speech: data.text };
```
Comment (obaid) [inline lib/providers/platforms/facebook.ts]: I think we should use constants to define types.

File: lib/providers/index.ts (hunk 1 above)
Comment (nonnontrivial) [reply lib/providers/index.ts]: Yes, this should be named better. `type` is the content block type, e.g. `jump`, `text`, etc.

File: lib/providers/platforms/facebook.ts (hunk 2 above)
Comment (nonnontrivial) [reply lib/providers/platforms/facebook.ts]: Yes, we don't even need to have the type determined from each platform class, we can just do this at the level of the provider class. The only exception to this is the google platform, which does not follow the numbering system all other platforms we support do.

**Pull request merged**